src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -15,39 +15,153 @@
+        append_path(&fx_dir, fx_name.c_str());
+
+        if (pal::directory_exists(fx_dir))
+        {
+            std::vector<pal::string_t> versions;
+            pal::readdir(fx_dir, &versions);
+            for (const auto& ver : versions)
+            {
+                // Make sure we filter out any non-version folders.
+                fx_ver_t parsed(-1, -1, -1);
+                if (fx_ver_t::parse(ver, &parsed, false))
+                {
+                    trace::verbose(_X("Found FX version [%s]"), ver.c_str());
+                    fx_versions->push_back(ver);
+                }
+            }
+        }
+    }
+}
+
+void get_all_sdk_versions(
+    const pal::string_t& own_dir,
+    std::vector<pal::string_t>* sdk_versions)
+{
+    trace::verbose(_X("Gathering SDK locations in [%s]"), own_dir.c_str());
+
+    auto sdk_dir = own_dir;
+    append_path(&sdk_dir, _X("sdk"));
+
+    if (pal::directory_exists(sdk_dir))
+    {
+        std::vector<pal::string_t> versions;
+        pal::readdir(sdk_dir, &versions);
+        for (const auto& ver : versions)
+        {
+            // Make sure we filter out any non-version folders.
+            fx_ver_t parsed(-1, -1, -1);
+            if (fx_ver_t::parse(ver, &parsed, false))
+            {
+                trace::verbose(_X("Found SDK version [%s]"), ver.c_str());
+                sdk_versions->push_back(ver);
+            }
+        }
+    }
+}
+
 /**
  * When the framework is not found, display detailed error message
  *   about available frameworks and installation of new framework.
  */
-void handle_missing_framework_error(const pal::string_t& fx_name, const pal::string_t& fx_version, const pal::string_t& fx_dir)
+void handle_missing_framework_error(
+    host_mode_t mode,
+    const pal::string_t& fx_name,
+    const pal::string_t& fx_version,
+    const pal::string_t& fx_dir,
+    const pal::string_t& own_dir)
 {
-    pal::string_t fx_ver_dirs = get_directory(fx_dir);
+    std::vector<pal::string_t> fx_versions;
+    pal::string_t fx_ver_dirs;
+    if (fx_dir.length())
+    {
+        fx_ver_dirs = fx_dir;
+        get_all_fx_versions(mode, get_directory(fx_dir), fx_name, &fx_versions);
+    }
+    else
+    {
+        fx_ver_dirs = own_dir;
+    }
+
+    get_all_fx_versions(mode, own_dir, fx_name, &fx_versions);
 
     // Display the error message about missing FX.
-    trace::error(_X("The specified framework '%s', version '%s' was not found."), fx_name.c_str(), fx_version.c_str());
+    if (fx_version.length())
+    {
+        trace::error(_X("The specified framework '%s', version '%s' was not found."), fx_name.c_str(), fx_version.c_str());
+    }
+    else
+    {
+        trace::error(_X("No framworks were found."));
+    }
+
     trace::error(_X("  - Check application dependencies and target a framework version installed at:"));
     trace::error(_X("      %s"), fx_ver_dirs.c_str());
+    trace::error(_X("  - The Dotnet framework can be installed from:"));
```

Comment:
Should this be `.NET` instead of `Dotnet`?

Reply:
I copied from the comment next to it, but I agree .NET should be used. Dotnet is used when naming directories and other cases where the 'dot' doesn't work